Projects/b_u585i_iot02a_ntz/Src/ota_pal/ota_pal_stm32u5_ntz.c

```diff
@@ -222,103 +378,163 @@ static HAL_StatusTypeDef prvFlashSetDualBankMode( void )
-                /* On successful option bytes loading the system should reset and control should not return from this function. */
-                status = HAL_FLASH_OB_Launch();
+                xObContext.USERConfig = OB_SWAP_BANK_DISABLE;
             }
 
+            xHalStatus = HAL_FLASHEx_OBProgram( &xObContext );
+
             ( void ) HAL_FLASH_OB_Lock();
+            ( void ) HAL_FLASH_Lock();
         }
 
-        ( void ) HAL_FLASH_Lock();
+        xResult = ( xHalStatus == HAL_OK ) ? pdTRUE : pdFALSE;
     }
 
-    return status;
+    return xResult;
 }
 
-
-static uint32_t prvGetActiveBank( void )
+static void prvOptionByteApply( void )
 {
-    HAL_StatusTypeDef status = HAL_ERROR;
-    uint32_t ulBank = 0UL;
+    vPetWatchdog();
 
-    status = HAL_FLASH_Unlock();
+    if( xTaskGetSchedulerState() == taskSCHEDULER_RUNNING )
+    {
+        vTaskSuspendAll();
+    }
 
-    if( status == HAL_OK )
+    LogSys( "Option Byte Launch in progress." );
+
+    /* Drain log buffers */
+    vDyingGasp();
+
+
+    ( void ) HAL_FLASH_Unlock();
+    ( void ) HAL_FLASH_OB_Unlock();
+    /* Generate System Reset to load the new option byte values ***************/
+    /* On successful option bytes loading the system should reset and control should not return from this function. */
+    ( void ) HAL_FLASH_OB_Launch();
+}
+
+static uint32_t prvGetActiveBank( void )
+{
+    if( ulBankAtBootup == 0 )
     {
-        status = HAL_FLASH_OB_Unlock();
+        ulBankAtBootup = prvGetBankSettingFromOB();
+    }
 
-        if( status == HAL_OK )
-        {
-            /* Get the Dual boot configuration status */
-            HAL_FLASHEx_OBGetConfig( &OBInit );
-            ulBank = ( ( OBInit.USERConfig & OB_SWAP_BANK_ENABLE ) == OB_SWAP_BANK_ENABLE ) ? FLASH_BANK_2 : FLASH_BANK_1;
+    return ulBankAtBootup;
+}
 
-            ( void ) HAL_FLASH_OB_Lock();
-        }
+static uint32_t prvGetBankSettingFromOB( void )
+{
+    uint32_t ulCurrentBank = 0UL;
+    FLASH_OBProgramInitTypeDef xObContext = { 0 };
 
-        ( void ) HAL_FLASH_Lock();
+    /* Get the Dual boot configuration status */
+    HAL_FLASHEx_OBGetConfig( &xObContext );
+
+    if( ( xObContext.USERConfig & OB_SWAP_BANK_ENABLE ) == OB_SWAP_BANK_ENABLE )
+    {
+        ulCurrentBank = FLASH_BANK_2;
+    }
+    else
+    {
+        ulCurrentBank = FLASH_BANK_1;
     }
 
-    return ulBank;
+    return ulCurrentBank;
 }
 
-static uint32_t prvGetPassiveBank( void )
+static uint32_t prvGetInactiveBank( void )
 {
-    uint32_t activeBank = prvGetActiveBank();
+    uint32_t ulActiveBank = prvGetActiveBank();
 
-    return ALTERNATE_BANK( activeBank );
+    uint32_t ulInactiveBank;
+
+    if( ulActiveBank == FLASH_BANK_1 )
+    {
+        ulInactiveBank = FLASH_BANK_2;
+    }
+    else if( ulActiveBank == FLASH_BANK_2 )
+    {
+        ulInactiveBank = FLASH_BANK_1;
+    }
+    else
+    {
+        ulInactiveBank = 0UL;
```

Comment:
This condition should probably be an assert. My concern is if `FLASH_BANK_1` definition could be is set to 0 then this fails.